File: Java-Core/src/main/java/presto/declaration/ConcreteCategoryDeclaration.java
package presto.declaration;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import presto.error.PrestoError;
import presto.error.SyntaxError;
import presto.grammar.CategoryMethodDeclarationList;
import presto.grammar.Identifier;
import presto.grammar.Operator;
import presto.runtime.Context;
import presto.runtime.Context.MethodDeclarationMap;
import presto.type.CategoryType;
import presto.type.IType;
import presto.utils.CodeWriter;
import presto.utils.IdentifierList;
import presto.value.ConcreteInstance;
import presto.value.IInstance;


public class ConcreteCategoryDeclaration extends CategoryDeclaration {
	
	IdentifierList derivedFrom;
	CategoryMethodDeclarationList methods;
	Map<String,IDeclaration> methodsMap = null;
	
	protected ConcreteCategoryDeclaration(Identifier name) {
		super(name);
	}
	
	public ConcreteCategoryDeclaration(Identifier name, IdentifierList attributes, 
			IdentifierList derivedFrom, CategoryMethodDeclarationList methods) {
		super(name, attributes);
		this.derivedFrom = derivedFrom;
		this.methods = methods!=null ? methods : new CategoryMethodDeclarationList();
	}

	@Override
	public IdentifierList getDerivedFrom() {
		return derivedFrom;
	}
	
	public CategoryMethodDeclarationList getMethods() {
		return methods;
	}
	
	@Override
	protected void toEDialect(CodeWriter writer) {
		boolean hasMethods = methods!=null && methods.size()>0;
		protoToEDialect(writer, hasMethods, false); // no mappings
		if(hasMethods)
			methodsToEDialect(writer, methods);
	}
	
	@Override
	protected void categoryTypeToEDialect(CodeWriter writer) {
		if(derivedFrom==null)
			writer.append("category");
		else
			derivedFrom.toDialect(writer, true);
	}
	

	@Override
	protected void toODialect(CodeWriter writer) {
		boolean hasMethods = methods!=null && methods.size()>0;
		toODialect(writer, hasMethods);
	}
	
	@Override
	protected void categoryTypeToODialect(CodeWriter writer) {
		writer.append("category");
	}
	
	@Override
	protected void categoryExtensionToODialect(CodeWriter writer) {
		if(derivedFrom!=null) {
			writer.append(" extends ");
			derivedFrom.toDialect(writer, true);
		}
	}
	
	@Override
	protected void bodyToODialect(CodeWriter writer) {
		for(IDeclaration decl : methods) {
			decl.toDialect(writer);
			writer.newLine();
		}
	}
	
	@Override
	protected void toPDialect(CodeWriter writer) {
		protoToPDialect(writer, derivedFrom);
		methodsToPDialect(writer);
	}
	
	@Override
	protected void categoryTypeToPDialect(CodeWriter writer) {
		writer.append("class");
	}

	private void methodsToPDialect(CodeWriter writer) {
		writer.indent();
		if(methods==null || methods.size()==0)
			writer.append("pass\n");
		else for(IDeclaration decl : methods) {
			decl.toDialect(writer);
			writer.newLine();
		}
		writer.dedent();
	}


	@Override
	public boolean hasAttribute(Context context, Identifier name) {
		if(super.hasAttribute(context, name))
			return true;
		if(hasDerivedAttribute(context,name))
			return true;
		return false;
	}
	
	private boolean hasDerivedAttribute(Context context, Identifier name) {
		if(derivedFrom==null)
			return false;
		for(Identifier ancestor : derivedFrom) {
			if(ancestorHasAttribute(ancestor,context,name))
				return true;
		}
		return false;
	}
		
	private static boolean ancestorHasAttribute(Identifier ancestor, Context context, Identifier name) {
		CategoryDeclaration actual = context.getRegisteredDeclaration(CategoryDeclaration.class, ancestor);
		if(actual==null)
			return false;
		return actual.hasAttribute(context, name);
	}

	@Override
	public IType check(Context context) throws SyntaxError {
		checkDerived(context);
		checkMethods(context);
		return super.check(context);
	}

	private void checkMethods(Context context) throws SyntaxError {
		if(methodsMap==null) {
			methodsMap = new HashMap<String,IDeclaration>();
			for(ICategoryMethodDeclaration method : methods)
				register(method,context);
		}
	}
			
			
	private void register(ICategoryMethodDeclaration method, Context context) throws SyntaxError {
		IDeclaration actual;
		if(method instanceof SetterMethodDeclaration) {
			actual = methodsMap.get("setter:"+method.getName().toString());
			if(actual!=null)
				throw new SyntaxError("Duplicate setter: \"" + method.getName().toString() + "\"");
			methodsMap.put("setter:"+method.getName().toString(),method);
		} else if(method instanceof GetterMethodDeclaration) {
			actual = methodsMap.get("getter:"+method.getName().toString());
			if(actual!=null)
				throw new SyntaxError("Duplicate getter: \"" + method.getName().toString() + "\"");
			methodsMap.put("getter:"+method.getName().toString(),method);
		} else {
			actual = methodsMap.get(method.getName().toString());
			if(actual==null) {
				actual = new MethodDeclarationMap(method.getName());
				methodsMap.put(method.getName().toString(), (MethodDeclarationMap)actual);
			}
			((MethodDeclarationMap)actual).register(method,context);
		}
		method.check(this,context);
	}

	private void checkDerived(Context context) throws SyntaxError {
		if(derivedFrom!=null) for(Identifier category : derivedFrom) {
			ConcreteCategoryDeclaration cd = context.getRegisteredDeclaration(ConcreteCategoryDeclaration.class, category);
			if(cd==null)
				throw new SyntaxError("Unknown category: \"" + category + "\"");
		}
	}

	@Override
	public boolean isDerivedFrom(Context context, CategoryType categoryType) {
		if(derivedFrom==null) 
			return false;
		for(Identifier ancestor : derivedFrom) {
			if(ancestor.equals(categoryType.getName()))
				return true;
			if(isAncestorDerivedFrom(ancestor,context,categoryType))
				return true;
		}
		return false;
	}

	private static boolean isAncestorDerivedFrom(Identifier ancestor, Context context, CategoryType categoryType) {
		IDeclaration actual = context.getRegisteredDeclaration(IDeclaration.class, ancestor);
		if(actual==null || !(actual instanceof CategoryDeclaration))
			return false;
		CategoryDeclaration cd = (CategoryDeclaration)actual;
		return cd.isDerivedFrom(context, categoryType);
	}
	
	@Override
	public IInstance newInstance() throws PrestoError {
		return new ConcreteInstance(this);
	}

	public GetterMethodDeclaration findGetter(Context context, Identifier attrName) throws SyntaxError {
		if(methodsMap==null)
			return null;
		IDeclaration method = methodsMap.get("getter:"+attrName); 
		if(method instanceof GetterMethodDeclaration)
			return (GetterMethodDeclaration)method;
		if(method!=null)
			throw new SyntaxError("Not a getter method!");
		return findDerivedGetter(context, attrName);
	}

	private GetterMethodDeclaration findDerivedGetter(Context context, Identifier attrName) throws SyntaxError {
		if(derivedFrom==null) 
			return null;
		for(Identifier ancestor : derivedFrom) {
			GetterMethodDeclaration method = findAncestorGetter(ancestor,context,attrName); 
			if(method!=null)
				return method;
		}
		return null;
	}

	private static GetterMethodDeclaration findAncestorGetter(Identifier ancestor, Context context, Identifier attrName) throws SyntaxError {
		IDeclaration actual = context.getRegisteredDeclaration(IDeclaration.class, ancestor);
		if(actual==null || !(actual instanceof ConcreteCategoryDeclaration))
			return null;
		ConcreteCategoryDeclaration cd = (ConcreteCategoryDeclaration)actual;
		return cd.findGetter(context, attrName);
	}

	public SetterMethodDeclaration findSetter(Context context, Identifier attrName) throws SyntaxError {
		if(methodsMap==null)
			return null;
		IDeclaration method = methodsMap.get("setter:"+attrName); 
		if(method instanceof SetterMethodDeclaration)
			return (SetterMethodDeclaration)method;
		if(method!=null)
			throw new SyntaxError("Not a setter method!");
		return findDerivedSetter(context,attrName);
	}

	private SetterMethodDeclaration findDerivedSetter(Context context, Identifier attrName) throws SyntaxError {
		if(derivedFrom==null) 
			return null;
		for(Identifier ancestor : derivedFrom) {
			SetterMethodDeclaration method = findAncestorSetter(ancestor,context,attrName); 
			if(method!=null)
				return method;
		}
		return null;
	}
	
	private static SetterMethodDeclaration findAncestorSetter(Identifier ancestor, Context context, Identifier attrName) throws SyntaxError {
		IDeclaration actual = context.getRegisteredDeclaration(IDeclaration.class, ancestor);
		if(actual==null || !(actual instanceof ConcreteCategoryDeclaration))
			return null;
		ConcreteCategoryDeclaration cd = (ConcreteCategoryDeclaration)actual;
		return cd.findSetter(context, attrName);
	}
	
	public Collection<IMethodDeclaration> findMemberMethods(Context context, Identifier name) throws SyntaxError {
		MethodDeclarationMap result = new MethodDeclarationMap(name);
		registerMemberMethods(context,result);
		return result.values();
	}
	
	
	private void registerMemberMethods(Context context, MethodDeclarationMap result) throws SyntaxError {
		registerThisMemberMethods(context,result);
		registerDerivedMemberMethods(context,result);
	}

	
	private void registerThisMemberMethods(Context context, MethodDeclarationMap result) throws SyntaxError {
		if(methodsMap==null)
			return;
		IDeclaration actual = methodsMap.get(result.getName().toString()); 
		if(actual==null)
			return;
		if(!(actual instanceof MethodDeclarationMap))
			throw new SyntaxError("Not a member method!");
		for(IMethodDeclaration method : ((MethodDeclarationMap)actual).values())
			result.registerIfMissing(method, context);
	}

	private void registerDerivedMemberMethods(Context context, MethodDeclarationMap result) throws SyntaxError {
		if(derivedFrom==null) 
			return;
		for(Identifier ancestor : derivedFrom)
			registerAncestorMemberMethods(ancestor,context,result); 
	}
	
	private void registerAncestorMemberMethods(Identifier ancestor, Context context, MethodDeclarationMap result) throws SyntaxError {
		IDeclaration actual = context.getRegisteredDeclaration(IDeclaration.class, ancestor);
		if(actual==null || !(actual instanceof ConcreteCategoryDeclaration))
			return;
		ConcreteCategoryDeclaration cd = (ConcreteCategoryDeclaration)actual;
		cd.registerMemberMethods(context, result);
	}

	public IMethodDeclaration findOperator(Context context, Operator operator, IType type) throws SyntaxError {
		Identifier methodName = new Identifier("operator_" + operator.name());
		Collection<IMethodDeclaration> methods = findMemberMethods(context, methodName);
		if(methods==null)
			return null;
		// find best candidate
		IMethodDeclaration candidate = null;
		for(IMethodDeclaration method : methods) {
			IType potential = method.getArguments().getFirst().getType(context);
			if(!type.isAssignableTo(context, potential))
				continue;
			if(candidate==null)
				candidate = method;
			else {
				IType currentBest = candidate.getArguments().getFirst().getType(context);
				if(currentBest.isAssignableTo(context, potential))
					candidate = method;
			}
		}
		return candidate;
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/declaration/ConcreteCategoryDeclaration.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var CategoryDeclaration = require("./CategoryDeclaration").CategoryDeclaration;
var SetterMethodDeclaration = require("./SetterMethodDeclaration").SetterMethodDeclaration;
var GetterMethodDeclaration = require("./GetterMethodDeclaration").GetterMethodDeclaration;
var MethodDeclarationMap = null;
var ConcreteInstance = require("../value/ConcreteInstance").ConcreteInstance;

exports.resolve = function() {
    MethodDeclarationMap = require("../runtime/Context").MethodDeclarationMap;
}

function ConcreteCategoryDeclaration(name, attributes, derivedFrom, methods) {
	CategoryDeclaration.call(this, name, attributes);
	this.derivedFrom = derivedFrom || null;
	this.methodsMap = null;
	this.methods = methods || [];
	return this;
}

ConcreteCategoryDeclaration.prototype = Object.create(CategoryDeclaration.prototype);
ConcreteCategoryDeclaration.prototype.constructor = ConcreteCategoryDeclaration;


ConcreteCategoryDeclaration.prototype.toEDialect = function(writer) {
    var hasMethods = this.methods!=null && this.methods.length>0;
    this.protoToEDialect(writer, hasMethods, false); // no bindings
    if(hasMethods)
        this.methodsToEDialect(writer, this.methods);
}

ConcreteCategoryDeclaration.prototype.categoryTypeToEDialect = function(writer) {
    if(this.derivedFrom==null)
        writer.append("category");
    else
        this.derivedFrom.toDialect(writer, true);
};


ConcreteCategoryDeclaration.prototype.toODialect = function(writer) {
    var hasMethods = this.methods!=null && this.methods.length>0;
    this.allToODialect(writer, hasMethods);
};

ConcreteCategoryDeclaration.prototype.categoryTypeToODialect = function(writer) {
    writer.append("category");
};


ConcreteCategoryDeclaration.prototype.categoryExtensionToODialect = function(writer) {
    if(this.derivedFrom!=null) {
        writer.append(" extends ");
        this.derivedFrom.toDialect(writer, true);
    }
};


ConcreteCategoryDeclaration.prototype.bodyToODialect = function(writer) {
    this.methodsToODialect (writer, this.methods);
};


ConcreteCategoryDeclaration.prototype.toSDialect = function(writer) {
    this.protoToSDialect(writer, this.derivedFrom);
    this.methodsToSDialect(writer);
};


ConcreteCategoryDeclaration.prototype.categoryTypeToSDialect = function(writer) {
    writer.append("class");
};


ConcreteCategoryDeclaration.prototype.methodsToSDialect = function(writer) {
    writer.indent();
    if(this.methods==null || this.methods.length==0)
        writer.append("pass\n");
    else for(var i=0;i<this.methods.length;i++) {
        var w = writer.newMemberWriter();
        this.methods[i].toDialect(w);
        writer.newLine();
    }
    writer.dedent();
};

ConcreteCategoryDeclaration.prototype.hasAttribute = function(context, name) {
	if (CategoryDeclaration.prototype.hasAttribute.call(this, context, name)) {
		return true;
	} else {
		return this.hasDerivedAttribute(context, name);
	}
};

ConcreteCategoryDeclaration.prototype.hasDerivedAttribute = function(context, name) {
	if(this.derivedFrom==null) {
		return false;
	}
	for(var i=0;i<this.derivedFrom.length;i++) {
		if(ConcreteCategoryDeclaration.ancestorHasAttribute(this.derivedFrom[i],context,name)) {
			return true;
		}
	}
	return false;
};

ConcreteCategoryDeclaration.ancestorHasAttribute = function(ancestor, context, name) {
	var actual = context.getRegisteredDeclaration(ancestor);
	if(!(actual instanceof CategoryDeclaration)) {
		return false;
	}
	return actual.hasAttribute(context, name);
};

ConcreteCategoryDeclaration.prototype.check = function(context) {
	this.checkDerived(context);
	this.checkMethods(context);
	return CategoryDeclaration.prototype.check.call(this, context);
}

ConcreteCategoryDeclaration.prototype.checkMethods = function(context) {
    this.registerMethods(context);
    for (var i = 0; i < this.methods.length; i++) {
        this.methods[i].memberCheck(this, context);
    }
};

ConcreteCategoryDeclaration.prototype.registerMethods = function(context) {
	if(this.methodsMap==null) {
        this.methodsMap = {};
		for(var i=0;i<this.methods.length;i++) {
            this.methods[i].memberOf = this;
			this.registerMethod(this.methods[i], context);
		}
	}
};


ConcreteCategoryDeclaration.prototype.registerMethod = function(method, context) {
	var actual;
	if(method instanceof SetterMethodDeclaration) {
		var key = "setter:"+method.name;
		actual = this.methodsMap[key] || null;
		if(actual!=null) {
			throw new SyntaxError("Duplicate setter: \"" + method.name + "\"");
		}
		this.methodsMap[key] = method;
	} else if(method instanceof GetterMethodDeclaration) {
		var key = "getter:"+method.name;
		actual = this.methodsMap[key] || null;
		if(actual!=null) {
			throw new SyntaxError("Duplicate getter: \"" + method.name + "\"");
		}
		this.methodsMap[key] = method;
	} else {
		var key = method.name;
		actual = this.methodsMap[key] || null;
		if(actual==null) {
			actual = new MethodDeclarationMap(method.name);
			this.methodsMap[key] = actual;
		}
		actual.register(method, context);
	}
};

ConcreteCategoryDeclaration.prototype.checkDerived = function(context) {
	if(this.derivedFrom!=null) {
		for(var i=0;i<this.derivedFrom.length;i++) {
			var cd = context.getRegisteredDeclaration(this.derivedFrom[i]) || null;
			if (cd == null) {
				throw new SyntaxError("Unknown category: \"" + this.derivedFrom[i] + "\"");
			}
		}
	}
};

ConcreteCategoryDeclaration.prototype.isDerivedFrom = function(context, categoryType) {
	if(this.derivedFrom==null) {
		return false;
	}
	for(var i=0;i<this.derivedFrom.length;i++) {
		var ancestor = this.derivedFrom[i];
		if(ancestor==categoryType.name) {
			return true;
		}
		if(ConcreteCategoryDeclaration.isAncestorDerivedFrom(ancestor,context,categoryType)) {
			return true;
		}
	}
	return false;
};

ConcreteCategoryDeclaration.isAncestorDerivedFrom = function(ancestor, context, categoryType) {
	var actual = context.getRegisteredDeclaration(ancestor);
	if(!(actual instanceof CategoryDeclaration)) {
		return false;
	}
	return actual.isDerivedFrom(context, categoryType);
};

ConcreteCategoryDeclaration.prototype.newInstance = function() {
	return new ConcreteInstance(this);
};

ConcreteCategoryDeclaration.prototype.findGetter = function(context, attrName) {
	if(this.methodsMap==null) {
		return null;
	}
	var method = this.methodsMap["getter:"+attrName] || null;
	if(method instanceof GetterMethodDeclaration) {
		return method;
	}
	if(method!=null) {
		throw new SyntaxError("Not a getter method!");
	}
	return this.findDerivedGetter(context, attrName);
};

ConcreteCategoryDeclaration.prototype.findDerivedGetter = function(context, attrName) {
	if(this.derivedFrom==null) {
		return null;
	}
	for(var i=0;i<this.derivedFrom;i++) {
		var method = ConcreteCategoryDeclaration.findAncestorGetter(this.derivedFrom[i],context,attrName);
		if(method!=null) {
			return method;
		}
	}
	return null;
}

ConcreteCategoryDeclaration.findAncestorGetter = function(ancestor, context, attrName) {
	var actual = context.getRegisteredDeclaration(ancestor);
	if(!(actual instanceof ConcreteCategoryDeclaration)) {
		return null;
	}
	return actual.findGetter(context, attrName);
};

ConcreteCategoryDeclaration.prototype.findSetter = function(context, attrName) {
	if(this.methodsMap==null) {
		return null;
	}
	var method = this.methodsMap["setter:"+attrName] || null;
	if(method instanceof SetterMethodDeclaration) {
		return method;
	}
	if(method!=null) {
		throw new SyntaxError("Not a setter method!");
	}
	return this.findDerivedSetter(context,attrName);
};

ConcreteCategoryDeclaration.prototype.findDerivedSetter = function(context, attrName) {
	if(this.derivedFrom==null) {
		return null;
	}
	for(var i=0;i<this.derivedFrom;i++) {
		var method = ConcreteCategoryDeclaration.findAncestorSetter(this.derivedFrom[i],context,attrName);
		if(method!=null) {
			return method;
		}
	}
	return null;
};

ConcreteCategoryDeclaration.findAncestorSetter = function(ancestor, context, attrName) {
	var actual = context.getRegisteredDeclaration(ancestor);
	if(!(actual instanceof ConcreteCategoryDeclaration)) {
		return null;
	}
	return actual.findSetter(context, attrName);
};


ConcreteCategoryDeclaration.prototype.findMemberMethods = function(context, name) {
	var result = new MethodDeclarationMap(name);
	this.registerMemberMethods(context,result);
	var names = Object.getOwnPropertyNames(result.methods);
	var list = [];
	for(var i=0;i<names.length;i++) {
		list.push(result.methods[names[i]]);
	}
	return list;
};


ConcreteCategoryDeclaration.prototype.registerMemberMethods = function(context, result) {
    this.registerMethods(context);
	this.registerThisMemberMethods(context,result);
	this.registerDerivedMemberMethods(context,result);
};


ConcreteCategoryDeclaration.prototype.registerThisMemberMethods = function(context, result) {
	if(this.methodsMap==null) {
		return;
	}
	var actual = this.methodsMap[result.name] || null;
	if(actual==null) {
		return;
	}
	if(!(actual instanceof MethodDeclarationMap)) {
		throw new SyntaxError("Not a member method!");
	}
	var names = Object.getOwnPropertyNames(actual.methods);
	for(var i=0;i<names.length;i++) {
		var method = actual.methods[names[i]];
		result.registerIfMissing(method, context);
	}
};

ConcreteCategoryDeclaration.prototype.registerDerivedMemberMethods = function(context, result) {
	if(this.derivedFrom==null) {
		return;
	}
	for(var i=0;i<this.derivedFrom.length;i++) {
		var ancestor = this.derivedFrom[i];
		this.registerAncestorMemberMethods(ancestor, context, result);
	}
};

ConcreteCategoryDeclaration.prototype.registerAncestorMemberMethods = function(ancestor, context, result) {
	var actual = context.getRegisteredDeclaration(ancestor);
	if(!(actual instanceof ConcreteCategoryDeclaration)) {
		return;
	}
	actual.registerMemberMethods(context, result);
};


ConcreteCategoryDeclaration.prototype.findOperator = function(context, operator, type) {
    var methodName = "operator_" + operator.name;
    var methods = this.findMemberMethods(context, methodName);
    if(methods==null)
        return null;
    // find best candidate
    var candidate = null;
    for(var i=0;i<methods.length;i++) {
        var method  = methods[i];
        var potential = method.args[0].getType(context);
        if(!type.isAssignableTo(context, potential))
            continue;
        if(candidate==null)
            candidate = method;
        else {
            var currentBest = candidate.args[0].getType(context);
            if(currentBest.isAssignableTo(context, potential))
                candidate = method;
        }
    }
    return candidate;
};

exports.ConcreteCategoryDeclaration = ConcreteCategoryDeclaration;
